internal/dataplane/kongstate/kongstate_test.go

```diff
@@ -1475,3 +1519,131 @@ func TestFillOverrides_ServiceFailures(t *testing.T) {
 		})
 	}
 }
+
+type fakeSchemaGetter struct {
+	schemas map[string]kong.Schema
+}
+
+var _ SchemaGetter = &fakeSchemaGetter{}
+
+func (s *fakeSchemaGetter) Get(_ context.Context, entityType string) (kong.Schema, error) {
+	schema, ok := s.schemas[entityType]
+	if !ok {
+		return nil, fmt.Errorf("schema not found")
+	}
+	return schema, nil
+}
+
+func TestKongState_FillCustomEntities(t *testing.T) {
+	customEntityTypeMeta := metav1.TypeMeta{
+		APIVersion: kongv1alpha1.GroupVersion.Group + "/" + kongv1alpha1.GroupVersion.Version,
+		Kind:       "KongCustomEntity",
+	}
+	testCases := []struct {
+		name                        string
+		initialState                *KongState
+		customEntities              []*kongv1alpha1.KongCustomEntity
+		plugins                     []*kongv1.KongPlugin
+		schemas                     map[string]kong.Schema
+		expectedCustomEntities      map[string][]custom.Object
+		expectedTranslationFailures map[k8stypes.NamespacedName]string
+	}{
+		{
+			name:         "single custom entity",
+			initialState: &KongState{},
+			customEntities: []*kongv1alpha1.KongCustomEntity{
+				{
+					TypeMeta: customEntityTypeMeta,
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "session-foo",
+					},
+					Spec: kongv1alpha1.KongCustomEntitySpec{
+						EntityType:     "sessions",
+						ControllerName: annotations.DefaultIngressClass,
+						Fields: apiextensionsv1.JSON{
+							Raw: []byte(`{"name":"session1"}`),
+						},
+					},
+				},
+			},
+			schemas: map[string]kong.Schema{
+				"sessions": {
+					"type":     "string",
+					"required": true,
+				},
+			},
+			expectedCustomEntities: map[string][]custom.Object{
+				"sessions": {
+					{
+						"name": "session1",
+					},
+				},
+			},
+		},
+		{
+			name:         "custom entity with unknown type",
+			initialState: &KongState{},
+			customEntities: []*kongv1alpha1.KongCustomEntity{
+				{
+					TypeMeta: customEntityTypeMeta,
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "session-foo",
+					},
+					Spec: kongv1alpha1.KongCustomEntitySpec{
+						EntityType:     "sessions",
+						ControllerName: annotations.DefaultIngressClass,
+						Fields: apiextensionsv1.JSON{
+							Raw: []byte(`{"name":"session1"}`),
+						},
+					},
+				},
+			},
+			expectedTranslationFailures: map[k8stypes.NamespacedName]string{
+				{
+					Namespace: "default",
+					Name:      "session-foo",
+				}: "failed to fetch entity schema for entity type sessions: schema not found",
+			},
+		},
+		// TODO: add cases including a custom entity referring to another service
```

Comment:
Can we solve this todo in this PR? If not let's create an issue and reference it here.

Reply:
It will be done in this PR later.

Reply:
Added the case.